src/picongpu/include/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp

```diff
@@ -0,0 +1,161 @@
+        HDINLINE float_X
+        operator()( const KinEnergyDensityType kinEnergyDensity, const DensityType density, ParticleType& parentIon, float_X randNr )
+        {
+
+            /* @TODO replace the float_64 with float_X and make sure the values are scaled to PIConGPU units */
+            float_64 constexpr protonNumber = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
+            float_64 constexpr neutronNumber = GetAtomicNumbers<ParticleType>::type::numberOfNeutrons;
+            float_64 chargeState = attribute::getChargeState(parentIon);
+
+            /* ionization condition */
+            if (chargeState < protonNumber)
+            {
+                /* atomic mass number (usually A) A = N + Z */
+                float_64 constexpr massNumber = neutronNumber + protonNumber;
+
+                /** @TODO replace the static_cast<float_64> by casts to float_X
+                 * or leave out entirely and compute everything in PIConGPU scaled units
+                 */
+                float_64 const densityUnit = static_cast<float_64>(particleToGrid::derivedAttributes::Density().getUnit()[0]);
+                float_64 const kinEnergyDensityUnit = static_cast<float_64>(particleToGrid::derivedAttributes::EnergyDensity().getUnit()[0]);
+                /* convert from kinetic energy density to average kinetic energy per particle */
+                float_64 const kinEnergyUnit = kinEnergyDensityUnit / densityUnit;
+                float_64 const kinEnergy = (kinEnergyDensity / density) * kinEnergyUnit;
+                /** convert kinetic energy in J to "temperature" in eV by assuming an ideal electron gas
+                 * E_kin = 3/2 k*T
+                 */
+                float_64 constexpr convKinEnergyToTemperature = UNITCONV_Joule_to_keV * float_64(1.e3) * float_64(2./3.);
+                float_64 const temperature = kinEnergy * convKinEnergyToTemperature;
+
+                float_64 const T_0 = temperature/math::pow(protonNumber,float_64(4./3.));
+
+                float_64 const T_F = T_0 / (float_64(1.) + T_0);
+
+                /* for all the fitting parameters @see ionizerConfig.param */
+
+                /** this is weird - I have to define temporary variables because
+                 * otherwise the math::pow function won't recognize those at the
+                 * exponent position */
+                float_64 constexpr TFA2_temp = TFA2;
+                float_64 constexpr TFA4_temp = TFA4;
+                float_64 constexpr TFBeta_temp = TFBeta;
+
+                float_64 const A = TFA1 * math::pow(T_0,TFA2_temp) + TFA3 * math::pow(T_0,TFA4_temp);
+
+                float_64 const B = -math::exp(TFB0 + TFB1*T_F + TFB2*math::pow(T_F,float_64(7.)));
+
+                float_64 const C = TFC1 * T_F + TFC2;
+
+                /* requires mass density in g/cm^3 */
+                float_64 constexpr nAvogadro = SI::N_AVOGADRO;
```

Comment:
please keep `constexpr` as the *first* argument. `constexpr` is *not* a type specifier which can be read right to left but is a specifier for the expression you are writing currently. discussed again today with @psychocoderHPC 

example difference to const: not all arguments on the right hand side need to be const to define a const.
but: to define a constexpr variable, the expression itself needs to be a constant expression.